src/metabase/query_processor/middleware/permissions.clj

```diff
@@ -108,11 +108,12 @@
     (let [card-id         (or *card-id* (:qp/source-card-id outer-query))
           required-perms  (query-perms/required-perms-for-query outer-query :already-preprocessed? true)
           source-card-ids (set/difference (:card-ids required-perms) (:card-ids gtap-perms))]
+      ;; On EE, check block permissions up front for all queries. If block perms are in place, reject all native queries
+      ;; (unless overriden by `gtap-perms`) and any queries that touch blocked tables/DBs
+      (check-block-permissions outer-query)
```

Comment:
I moved this up since I think we should do the explicit block perm check unconditionally, not just for saved cards. It results in a redundancy since `query-perms/check-data-perms` will also reject the same queries for not having the minimum required data perms. But I think this approach is safer and avoids confusing cases like the original issue where there's a different behavior before and after saving a card. I can change it back if anyone disagrees.